modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java

```diff
@@ -358,11 +358,22 @@ public void validateAggregateParams(SqlCall aggCall,
     public RelDataType deriveType(SqlValidatorScope scope, SqlNode expr) {
         RelDataType dataType = super.deriveType(scope, expr);
 
+        // If type of dynamic parameter has not been inferred, use a type of its value.
```

Comment:
@zstan I think we should check whether `expr` is a cast expr, and if it `true`, and then validate its arguments ensure that it is legal cast. Because at the moment it looks like the check is applied to dynamic parameters in any expression or maybe I am missing something here?

Reply:
make sense, done.